typing/typecore.ml

```diff
@@ -4783,85 +4783,87 @@ let type_clash_of_trace trace =
-  | Some (Tpat_constant const) -> report_literal_type_constraint ppf const diff
-  | _ -> ()
+  | Some (Tpat_constant const) -> report_literal_type_constraint const diff
+  | _ -> []
 
 (* Hint when using int operators (eg. `+`)
    on other kind of integer and floats *)
 let report_numeric_operator_clash_hints actual_type operator =
   let stdlib = Path.Pident (Ident.create_persistent "Stdlib") in
   let stdlib_qualified mod_ val_ = Path.Pdot (Path.Pdot (stdlib, mod_), val_) in
-  let hint expected_op =
-    Some (fun ppf ->
-      fprintf ppf "@[Hint:@ Did you mean to use `%a'?@]"
-        Printtyp.path expected_op
-    )
-  in
-  let hint ~add ~sub ~mul ~div ~mod_ () =
-    let is_op op = Path.same operator (Path.Pdot (stdlib, op)) in
-    if is_op "+" then hint add
-    else if is_op "-" then hint sub
-    else if is_op "*" then hint mul
-    else if is_op "/" then hint div
-    else if is_op "mod" then hint mod_
-    else None
-  in
-  let hint_qualified name =
+  let is_op op = Path.same operator (Path.Pdot (stdlib, op)) in
+  let expecting_qualified name =
     let qualified = stdlib_qualified name in
-    hint ~add:(qualified "add") ~sub:(qualified "sub") ~mul:(qualified "mul")
-      ~div:(qualified "div") ~mod_:(qualified "rem") ()
+    if is_op "+" then Some (qualified "add")
+    else if is_op "-" then Some (qualified "sub")
+    else if is_op "*" then Some (qualified "mul")
+    else if is_op "/" then Some (qualified "div")
+    else if is_op "mod" then Some (qualified "rem")
+    else None
   in
-  let hint_std () =
+  let expecting_float () =
     let qualified id = Path.Pdot (stdlib, id) in
-    hint ~add:(qualified "+.") ~sub:(qualified "-.") ~mul:(qualified "*.")
-      ~div:(qualified "/.") ~mod_:(stdlib_qualified "Float" "rem") ()
+    if is_op "+" then Some (qualified "+.")
+    else if is_op "-" then Some (qualified "-.")
+    else if is_op "*" then Some (qualified "*.")
+    else if is_op "/" then Some (qualified "/.")
+    else if is_op "mod" then Some (stdlib_qualified "Float" "rem")
+    else None
   in
-  let expecting = Path.same actual_type in
-  if expecting Predef.path_int32 then
-    hint_qualified "Int32"
-  else if expecting Predef.path_int64 then
-    hint_qualified "Int64"
-  else if expecting Predef.path_nativeint then
-    hint_qualified "Nativeint"
-  else if expecting Predef.path_float then
-    hint_std ()
-  else None
+  let expecting_op =
+    if Path.same actual_type Predef.path_int32 then
+      expecting_qualified "Int32"
+    else if Path.same actual_type Predef.path_int64 then
+      expecting_qualified "Int64"
+    else if Path.same actual_type Predef.path_nativeint then
+      expecting_qualified "Nativeint"
+    else if Path.same actual_type Predef.path_float then
+      expecting_float ()
+    else None
+  in
+  match expecting_op with
+  | Some op ->
+      Some (fun ppf ->
+        fprintf ppf "@[Hint:@ Did you mean to use `%a'?@]"
+          Printtyp.path op
+      )
+  | None -> None
```

Comment:
I think we're almost good. Since you refactored this function as well, I think for symmetry it would be nicer to make it return a `Location.msg list` directly. You would pass an extra `~loc` argument to this function, and call `Location.msg` to create the message instead of `fprintf` and finally creating the message later in `report_application_clash_hints`.

Reply:
I changed that,